libminifi/test/unit/CronTests.cpp

```diff
@@ -0,0 +1,686 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string>
+
+#include "../Catch.h"
+#include "utils/Cron.h"
+#include "date/date.h"
+#include "date/tz.h"
+
+using std::chrono::system_clock;
+using std::chrono::seconds;
+using org::apache::nifi::minifi::utils::Cron;
+
+
+void checkNext(const std::string& expr, const date::zoned_time<seconds>& from, const date::zoned_time<seconds>& next) {
+  auto cron_expression = Cron(expr);
+  auto next_trigger = cron_expression.calculateNextTrigger(from.get_local_time());
+  CHECK(next_trigger == next.get_local_time());
+}
+
+
+TEST_CASE("Cron expression ctor tests", "[cron]") {
+  REQUIRE_THROWS(Cron("1600 ms"));
+  REQUIRE_THROWS(Cron("foo"));
+  REQUIRE_THROWS(Cron("61 0 0 * * *"));
+  REQUIRE_THROWS(Cron("0 61 0 * * *"));
+  REQUIRE_THROWS(Cron("0 0 24 * * *"));
+  REQUIRE_THROWS(Cron("0 0 0 32 * *"));
+
+  REQUIRE_THROWS(Cron("1banana * * * * * *"));
+  REQUIRE_THROWS(Cron("* 1banana * * * * *"));
+  REQUIRE_THROWS(Cron("* * 1banana * * * *"));
+  REQUIRE_THROWS(Cron("* * * 1banana * * *"));
+  REQUIRE_THROWS(Cron("* * * * 1banana * *"));
+  REQUIRE_THROWS(Cron("* * * * DECbanana * *"));
+  REQUIRE_THROWS(Cron("* * * * * WEDbanana *"));
+
+  REQUIRE_THROWS(Cron("* * * * * * 1banana"));
+  REQUIRE_THROWS(Cron("* * * * * * 2000banana"));
+
+  REQUIRE_THROWS(Cron("1G * * * * * *"));
+  REQUIRE_THROWS(Cron("* 1G * * * * *"));
+  REQUIRE_THROWS(Cron("* * 1G * * * *"));
+  REQUIRE_THROWS(Cron("* * * 1G * * *"));
+  REQUIRE_THROWS(Cron("* * * * 1G * *"));
+  REQUIRE_THROWS(Cron("* * * * * 1G *"));
+  REQUIRE_THROWS(Cron("* * * * * * 1G"));
+
+  // Number of fields must be 6 or 7
+  REQUIRE_THROWS(Cron("* * * * *"));
+  REQUIRE_NOTHROW(Cron("* * * * * *"));
+  REQUIRE_NOTHROW(Cron("* * * * * * *"));
+  REQUIRE_THROWS(Cron("* * * * * * * *"));
+
+  // LW can only be used in 4th field
+  REQUIRE_THROWS(Cron("LW * * * * * *"));
+  REQUIRE_THROWS(Cron("* LW * * * * *"));
+  REQUIRE_THROWS(Cron("* * LW * * * *"));
+  REQUIRE_NOTHROW(Cron("* * * LW * * *"));
+  REQUIRE_THROWS(Cron("* * * * LW * *"));
+  REQUIRE_THROWS(Cron("* * * * * LW *"));
+  REQUIRE_THROWS(Cron("* * * * * * LW"));
+
+  // n#m can only be used in 6th field
+  REQUIRE_THROWS(Cron("2#1 * * * * * *"));
+  REQUIRE_THROWS(Cron("* 2#1 * * * * *"));
+  REQUIRE_THROWS(Cron("* * 2#1 * * * *"));
+  REQUIRE_THROWS(Cron("* * * 2#1 * * *"));
+  REQUIRE_THROWS(Cron("* * * * 2#1 * *"));
+  REQUIRE_NOTHROW(Cron("* * * * * 2#1 *"));
+  REQUIRE_THROWS(Cron("* * * * * * 2#1"));
+
+  // L can only be used in 4th, 5th, 6th fields
+  REQUIRE_THROWS(Cron("L * * * * * *"));
+  REQUIRE_THROWS(Cron("* L * * * * *"));
+  REQUIRE_THROWS(Cron("* * L * * * *"));
+  REQUIRE_NOTHROW(Cron("* * * L * * *"));
+  REQUIRE_THROWS(Cron("* * * * L * *"));
+  REQUIRE_NOTHROW(Cron("* * * * * L *"));
+  REQUIRE_THROWS(Cron("* * * * * * L"));
```

Comment:
The comment says 4,5,6, but the code says 4,6.

Reply:
Good catch, dont know why the comment lies :+1: fortunately the code was right not the comment.
https://github.com/apache/nifi-minifi-cpp/pull/1335/commits/a7db03ebb2a795bff86aff210e4e29bcf424c372#diff-99182a0f9eeb09751e25ef7c3291f861bbcafcdfb35bfd7adfb666d075aff57eR86